services/shhext/chat/encryption_multi_device_test.go

```diff
@@ -152,25 +142,107 @@ func (s *EncryptionServiceMultiDeviceSuite) TestProcessPublicBundleOutOfOrder()
 	s.Require().NoError(err)
 
 	// Alice1 creates a bundle
-	alice1Bundle, err := s.alice1.CreateBundle(aliceKey)
+	alice1Bundle, err := s.services[aliceUser].encryptionServices[0].CreateBundle(aliceKey)
 	s.Require().NoError(err)
 
 	// Alice2 Receives the bundle
-	_, err = s.alice2.ProcessPublicBundle(aliceKey, alice1Bundle)
+	_, err = s.services[aliceUser].encryptionServices[1].ProcessPublicBundle(aliceKey, alice1Bundle)
 	s.Require().NoError(err)
 
 	// Alice2 Creates a Bundle
-	_, err = s.alice2.CreateBundle(aliceKey)
+	_, err = s.services[aliceUser].encryptionServices[1].CreateBundle(aliceKey)
 	s.Require().NoError(err)
 
 	// We enable the installation
-	err = s.alice2.EnableInstallation(&aliceKey.PublicKey, "alice1")
+	err = s.services[aliceUser].encryptionServices[1].EnableInstallation(&aliceKey.PublicKey, "alice1")
 	s.Require().NoError(err)
 
 	// It should contain both bundles
-	alice2MergedBundle1, err := s.alice2.CreateBundle(aliceKey)
+	alice2MergedBundle1, err := s.services[aliceUser].encryptionServices[1].CreateBundle(aliceKey)
 	s.Require().NoError(err)
 
 	s.Require().NotNil(alice2MergedBundle1.GetSignedPreKeys()["alice1"])
 	s.Require().NotNil(alice2MergedBundle1.GetSignedPreKeys()["alice2"])
 }
+
+func pairDevices(s *serviceAndKey, target int) error {
+	device := s.encryptionServices[target]
+	for i := 0; i < len(s.encryptionServices); i++ {
+		b, err := s.encryptionServices[i].CreateBundle(s.key)
+
+		if err != nil {
+			return err
+		}
+
+		_, err = device.ProcessPublicBundle(s.key, b)
+		if err != nil {
+			return err
+		}
+
+		err = device.EnableInstallation(&s.key.PublicKey, s.encryptionServices[i].config.InstallationID)
+		if err != nil {
+			return nil
+		}
+	}
+	return nil
+}
+
+func (s *EncryptionServiceMultiDeviceSuite) TestMaxDevices() {
+	err := pairDevices(s.services[aliceUser], 0)
+	s.Require().NoError(err)
+	alice1 := s.services[aliceUser].encryptionServices[0]
+	bob1 := s.services[bobUser].encryptionServices[0]
+	aliceKey := s.services[aliceUser].key
+	bobKey := s.services[bobUser].key
+
+	// Check bundle is ok
+	// No installation is enabled
+	aliceBundle, err := alice1.CreateBundle(aliceKey)
+	s.Require().NoError(err)
+
+	// Check all installations are correctly working, and that the oldest device is not there
+	preKeys := aliceBundle.GetSignedPreKeys()
+	s.Require().Equal(3, len(preKeys))
+	s.Require().NotNil(preKeys["alice1"])
```

Comment:
What about `s.Require().NotNil(preKeys["alice2"])`?

Reply:
`alice2` should not be there, we only ever propagate `maxInstallations` (in this case is set to 3):
`alice1` is the current device, so it is always included
`alice2` is the oldest device, so it is rotated out, while `alice3`/`alice4` are kept as newer
I could add  `s.Require().Nil(preKeys["alice2"])` to make it more explicit (we count the devices so we know `alice2` is not there)

Reply:
I missed that. I think that would be helpful with a comment why `alice2` is not there 👍

Reply:
yes, makes sense, already added!